File: ios/BackgroundImageLoader.swift
//
//  BackgroundImageLoader.swift
//  VisionCameraBackgroundFilter
//
//  Background decoding, shared by the live renderer and the offline bake.
//
//  Both paths MUST load a background the same way, or the preview and the delivered
//  file disagree about the very pixels this feature exists to composite. Keeping the
//  decode (and its downscale) in one place is part of the same argument as
//  `BackgroundGeometry`: two implementations, spec + native, never one per path.
//

import CoreImage
import Foundation

enum BackgroundImageLoader {

  /**
   * Longest edge a decoded background may keep.
   *
   * Mirrors `MAX_BACKGROUND_EDGE_PX` in `src/assets/background/index.ts`.
   * Duplicated because native cannot read the TS constant; keep them in step.
   */
  static let maxEdgePx: CGFloat = 1920

  /**
   * Decodes a background from the URI forms the JS side can produce: a `file://` or
   * absolute path in release, and an `http://` Metro URL in dev.
   *
   * Returns nil on any failure — callers then draw the unmodified frame rather than
   * a broken composite, and never log.
   */
  static func load(uri: String) -> CIImage? {
    let image: CIImage?
    if uri.hasPrefix("http://") || uri.hasPrefix("https://") {
      guard let url = URL(string: uri),
            let data = try? Data(contentsOf: url)
      else { return nil }
      image = CIImage(data: data)
    } else {
      let url = uri.hasPrefix("file://") ? URL(string: uri) : URL(fileURLWithPath: uri)
      guard let url else { return nil }
      image = CIImage(contentsOf: url)
    }

    guard let image else { return nil }
    return downscaleIfNeeded(image)
  }

  /**
   * Scales an oversized background down once, at load.
   *
   * The bundled images reach 8192x5464. CoreImage is lazy enough to survive that,
   * but every render pass would then sample a texture ~18x larger than the frame it
   * composites into — per frame, for every frame of an offline bake.
   */
  static func downscaleIfNeeded(_ image: CIImage) -> CIImage {
    let extent = image.extent
    guard extent.width > 0, extent.height > 0 else { return image }
    let longest = max(extent.width, extent.height)
    guard longest > maxEdgePx else { return image }
    let scale = maxEdgePx / longest
    return image.transformed(by: CGAffineTransform(scaleX: scale, y: scale))
  }
}

File: android/src/main/java/com/saigontechnology/backgroundfilter/BackgroundImageLoader.kt
package com.saigontechnology.backgroundfilter

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.margelo.nitro.NitroModules
import java.io.File
import java.net.URI
import java.net.URL

/**
 * Background decoding, shared by the live renderer and the offline bake.
 *
 * The Android counterpart of `ios/BackgroundImageLoader.swift`, and it exists for
 * the same reason: both paths MUST load a background the same way, or the preview
 * and the delivered file disagree about the very pixels this feature exists to
 * composite. The two used to carry a private copy of this each, which is how the
 * resource branch below came to be missing from both.
 */
object BackgroundImageLoader {

  /**
   * Longest edge a decoded background is allowed to keep, in pixels.
   *
   * Mirrors `MAX_BACKGROUND_EDGE_PX` in `src/assets/background/index.ts`.
   * Duplicated because native cannot read the TS constant; keep them in step.
   *
   * The downscale is not an optimisation, it is a correctness requirement: an
   * 8192x5464 source decodes to ~171 MB of RGBA — an OOM on mid-range devices —
   * and exceeds the 4096px `GL_MAX_TEXTURE_SIZE` many GPUs report, which makes the
   * texture upload fail outright. Consumers can inject arbitrary images, so this
   * cannot rely on assets being well-sized.
   */
  const val MAX_BACKGROUND_EDGE_PX = 1920

  /**
   * Decodes a background from any URI form the JS side can produce.
   *
   * Returns null on any failure — callers then draw the camera frame unmodified
   * rather than a broken composite, and never log.
   */
  fun load(uri: String): Bitmap? =
    runCatching {
      val bytes = readBytes(uri) ?: return null

      // Two passes: measure, then decode subsampled. `inSampleSize` only honours
      // powers of two, so this lands at or below the cap, never above it.
      val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
      BitmapFactory.decodeByteArray(bytes, 0, bytes.size, bounds)

      var sampleSize = 1
      while (maxOf(bounds.outWidth, bounds.outHeight) / sampleSize > MAX_BACKGROUND_EDGE_PX) {
        sampleSize *= 2
      }

      BitmapFactory.decodeByteArray(
        bytes,
        0,
        bytes.size,
        BitmapFactory.Options().apply { inSampleSize = sampleSize },
      )
    }.getOrNull()

  /**
   * Reads the bytes behind a background URI.
   *
   * The forms `Image.resolveAssetSource` can hand back for a `require()`d asset,
   * plus the `{ uri }` sources a consumer may inject:
   *
   * | Form                                  | Where it comes from                    |
   * | ------------------------------------- | -------------------------------------- |
   * | `http://10.0.2.2:8081/assets/…`       | Metro, in a dev build                  |
   * | `file:///…` / `/data/…`               | an injected `{ uri }`, a cached file   |
   * | `assets_…_bgoffice` (no scheme, no /) | **a RELEASE build's drawable resource** |
   *
   * That last row is the one that used to be missing, and its absence is invisible
   * in every dev build. React Native packages `require()`d images into
   * `res/drawable-*` for release Android builds and resolves them by RESOURCE
   * IDENTIFIER — `AssetSourceResolver.resourceIdentifierWithoutScale`, e.g.
   * `packages_visioncamerabackgroundfilter_src_assets_background_bgoffice`. There is
   * no file at that path, so `File(uri).readBytes()` threw, `load` swallowed it, and
   * the renderer composited a null background: the filter silently did nothing on
   * every release APK, in the preview AND in the offline bake, while a dev build
   * (Metro serves an `http://` URL) worked perfectly. iOS is unaffected — its
   * release assets stay real files in the bundle.
   */
  private fun readBytes(uri: String): ByteArray? =
    when {
      uri.startsWith("http://") || uri.startsWith("https://") ->
        URL(uri).openStream().use { it.readBytes() }

      uri.startsWith("file://") -> File(URI(uri)).readBytes()

      else -> {
        // A path is tried first so an injected absolute path keeps working exactly
        // as it did; the resource lookup is a fallback, not a replacement.
        val file = File(uri)
        if (file.isFile) file.readBytes() else readDrawableResource(uri)
      }
    }

  /**
   * Reads a drawable resource by name, as the JS asset registry addresses it.
   *
   * `openRawResource` returns the packaged file's own bytes — AAPT stores a jpg/png
   * in `res/drawable-*` verbatim rather than compiling it — so this feeds
   * `BitmapFactory` the same bytes every other branch does, and the density
   * qualifier is resolved by the resource system.
   */
  private fun readDrawableResource(name: String): ByteArray? {
    val context: Context = NitroModules.applicationContext ?: return null
    val resources = context.resources
    val id = resources.getIdentifier(name, "drawable", context.packageName)
    if (id == 0) return null
    return resources.openRawResource(id).use { it.readBytes() }
  }
}
